tiledb/sm/buffer/buffer.h

```diff
@@ -418,6 +435,226 @@ class ConstBuffer : public BufferBase {
+      : buffer_owner_(other.buffer_owner_)
+      // If we are copying an owned buffer, we must update buffer_ to point to
+      // the new buffer_owner_, otherwise we keep the existing buffer_.
+      , buffer_(other.is_owned() ? buffer_owner_ : other.buffer_) {
+  }
+
+  /** Default copy assignment operator. */
+  SerializationBuffer& operator=(const SerializationBuffer& other) {
+    buffer_owner_ = other.buffer_owner_;
+    buffer_ = other.is_owned() ? buffer_owner_ : other.buffer_;
+    return *this;
+  }
+
+  /** Default move constructor. */
+  // Use of noexcept is important to avoid copies when BufferList is being
+  // resized.
+  SerializationBuffer(SerializationBuffer&& other) noexcept
+      : buffer_owner_(std::move(other.buffer_owner_))
+      // Invalidate the buffer_ from other to avoid use-after-move bugs.
+      , buffer_(std::exchange(other.buffer_, {})) {
+  }
+
+  /** Default move assignment operator. */
+  SerializationBuffer& operator=(SerializationBuffer&& other) {
+    buffer_owner_ = std::move(other.buffer_owner_);
+    buffer_ = std::exchange(other.buffer_, {});
+    return *this;
+  }
+
+  /** Allocator-aware copy constructor. */
+  SerializationBuffer(
+      const SerializationBuffer& other, const allocator_type& alloc)
+      : buffer_owner_(other.buffer_owner_, alloc)
+      , buffer_(other.is_owned() ? buffer_owner_ : other.buffer_) {
+  }
+
+  /** Allocator-aware move constructor. */
+  SerializationBuffer(SerializationBuffer&& other, const allocator_type& alloc)
+      : buffer_owner_(std::move(other.buffer_owner_), alloc)
+      // We can't copy the buffer_ from other because the buffer might be
+      // reallocated if the allocators are different.
+      , buffer_(other.is_owned() ? buffer_owner_ : other.buffer_) {
+    other.buffer_ = {};
+  }
+
+  /**
+   * Constructor for an owned buffer of a given size.
+   *
+   * The data is intended to be modified later with the owned_mutable_span()
+   * function.
+   *
+   * @param size The size of the buffer.
+   * @param alloc Optional allocator for the buffer.
+   */
+  SerializationBuffer(size_t size, const allocator_type& alloc = {})
+      : buffer_owner_(size, alloc)
+      , buffer_(buffer_owner_) {
+  }
+
+  /**
+   * Shorthand constructor for a non-owned buffer to a given pointer-size pair.
+   *
+   * @param data The data of the buffer.
+   * @param size The size of the buffer.
+   */
+  SerializationBuffer(
+      NonOwnedMarker,
+      const void* data,
+      size_t size,
+      const allocator_type& alloc = {})
+      : buffer_owner_(alloc)
+      , buffer_(static_cast<const char*>(data), size) {
+  }
+
+  /**
+   * Returns the buffer's allocator. This is required to make the type
+   * allocator-aware.
+   */
+  inline allocator_type get_allocator() const noexcept {
+    return buffer_owner_.get_allocator();
+  }
+
+  bool is_owned() const {
+    return buffer_.data() == buffer_owner_.data();
+  }
```

Comment:
On second thought I could make this private. It has no use cases outside of this class. It could be used to guard for calling `owned_mutable_span`, but you are supposed to call this function on buffers you created yourself, and you know whether they are owned.

And reminder for me to document it.